src/l1_fetcher.rs

```diff
@@ -0,0 +1,123 @@
+use ethers::{abi::Contract, prelude::*, providers::Provider};
+use eyre::Result;
+use state_reconstruct::parse_calldata;
+use state_reconstruct::CommitBlockInfoV1;
+use state_reconstruct::BLOCK_STEP;
+use state_reconstruct::GENESIS_BLOCK;
+use state_reconstruct::ZK_SYNC_ADDR;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+
+pub struct L1Fetcher {
+    provider: Provider<Http>,
+    contract: Contract,
+}
+
+impl L1Fetcher {
+    pub fn new(http_url: &str) -> Result<Self> {
```

Comment:
I'll continue improving this if the general direction is right. There are parameters missing (e.g. BLOCK_SKIP) now and I'm thinking if we should also move the relevant `lib.rs` constants here?

Reply:
We should probably consider having a dedicated `constants.rs` file instead I think.